### ISSUE TITLE
Docker buffering documentation

### ISSUE THREAD

**Issue by jaygorrell** [fluent-plugins-nursery/fluent-plugin-concat#42]
#### Problem
I think this project is going to have a popular use-case as more people move to Docker 1.13 (Kubernetes just upgraded), which added 16kb line limits, breaking logs up if they exceed that limit. Many people log "single-line json" in Docker so the end result here is that a "\n" signals the end if the real line.

Would it be possible to get a working solution for this? Ideally documented but I'd be happy with one in an issue here too. :)

Using the standard in the test files:
```
messages = [
        { "host" => "example.com", "message" => "{\"key1\": \"value1\",\"key2\": \"value2\"}\n" },
        { "host" => "example.com", "message" => "{\"key3\": \"value3\",\"key4\": \"value4\"," },
        { "host" => "example.com", "message" => "\"key5\": \"value5\",\"key6\": \"value6\"," },
        { "host" => "example.com", "message" => "\"key7\": \"value7\",\"key8\": \"value8\"}\n" },
        { "host" => "example.com", "message" => "{\"key9\": \"value9\",\"key0\": \"value0\"," },
        { "host" => "example.com", "message" => "\"key1\": \"value1\",\"key2\": \"value2\"}\n" },
]
```
The expected output would be something like...
```
{ "host" => "example.com","message" => "{\"key1\": \"value1\",\"key2\": \"value2\"}\n" }
{ "host" => "example.com","message" => "{\"key3\": \"value3\",\"key4\": \"value4\",\"key5\": \"value5\",\"key6\": \"value6\",\"key7\": \"value7\",\"key8\": \"value8\"}\n" },
{ "host" => "example.com","message" => "{\"key9\": \"value9\",\"key0\": \"value0\",\"key1\": \"value1\",\"key2\": \"value2\"}\n" },
```

This should be fairly straightforward but I get weird results where the first grouping comes out correctly but every line after that is on it's own and not concatenated. I imagine I'm missing one of the other regex attributes that is needed.

For what it's worth, I'm using this for regex start:
`/.*(?<!\\n)/`

**Comment by jaygorrell** [comment]
As a follow-up to make this test-case a bit easier, I've been trying to tweak the built-in tests to fit this issue but even a fairly basic setup isn't working. I'm likely doing something wrong but it appears I can't get things to come out correctly by matching the end of a line.

I'm basically working with `test_multiline_start_end_regexp` and trying to modify it so that the start regex is any line that doesn't end with `end` and the end regex is lines that end with `end`. The test messages and expected results should be the exact same -- only the regex changes.

Here is what I would have expected to work:
```
multiline_start_regexp /.*(?<!end)$/
multiline_end_regexp /.*(?<=end)$/
```

Oddly, the result of this is 
```
{"container_id"=>"1", "message"=>"start message end"},
{"container_id"=>"1", "message"=>"start"},
{"container_id"=>"1", "message"=>" message1"},
{"container_id"=>"1", "message"=>" message2\n" + "end"}
```

I'm not even sure why `message2` triggered the start of a new record to concatenate but the previous two lines did not.

**Comment by okkez** [comment]
@jaygorrell  Could you try v2.2.0?

**Comment by jaygorrell** [comment]
Thanks @okkez, I'm in contact with @denniswebb and that certainly helped a lot. I'm also returning early on line 87 if `key` isn't in the current record - this prevents dropping logs that aren't needed to go through the concat process.

In addition (for anyone else hitting this), I needed to extract the docker container's file path (via tag) to use as the stream id -- without this I was getting lots of unrelated logs concatenated with each other:
```
<filter containers.**>
  @type record_transformer
  <record>
    stream_id ${tag}
  </record>
</filter>
```

Doing this and using `stream_id` for the `stream_identity_key` really helped a lot. I still have some logs incorrectly concatenated but at least they are from the same service. I could probably pass things through a timestamp-based sort if needed to optimize this further.

**Comment by okkez** [comment]
@jaygorrell Could you show me logs to reproduce your situation?
I've checked #45 but I could not reproduce the situation.

**Comment by jaygorrell** [comment]
Added a test. You can see that removing the check causes it to fail... perhaps that is intended but I think for most people, they wouldn't want to just drop those logs.

**Comment by okkez** [comment]
Sorry for delay.
Released v2.2.1.
